### DISCUSSION TITLE
Messages (Handler/Sender): Refactor Packet Handling and Sending into a DDD approach.

### DISCUSSION THREAD

**Discussion by ovflowd** [NoCode-NoLife/melia#183] [Research]
Right now all the Messages handled by a Channel, both to Handle packets coming from the Client-Side and also to Send Packets to the Server-Side are being handled by two files:

- `Network/ChannelPacketHandler`
- `Network/ChannelPacketSender`

Those two files are already enormously big and already might impact the maintainability of the Melia Core.

The suggestion is to first move the part of the code where messages are being handled and written to a different part of the Emulator. (It is important to mention to leave all-kind of Business-Logic and Application-Logic outside the scope of composing the bodies and reading the bodies of packets.

This means that the refactored methods for handling and sending packets (The part that composes the bodies of the messages and the part that reads them) should be isolated.

### Sender-Reader-Handler Approach

* A reader should only contain pieces of code to read the Packet and return the data that was read as a return value of the method
* A Sender should only contain pieces of code to write the bodies of the packets and return the Buffer ready to be sent.
* A handler should invoke a Reader function when it wants to read a packet and Invoke a Sender function when it wants to send something.

#### An example of this strategy would be:

```csharp

private readonly CharacterSessionReader Reader = new CharacterSessionReader();
private readonly CharacterSessionSender Sender = new CharacterSessionSender();

[PacketHandler(Op.CZ_LOGOUT)]
public void CZ_LOGOUT(ChannelConnection conn, Packet packet)
{
    var { extra, unkByte } = Reader.CZ_LOGOUT();

    Log.Info("User '{0}' is logging out.", conn.Account.Name);

	Send(Sender.ZC_SAVE_INFO(conn));
	Send(Sender.ZC_LOGOUT_OK(conn));
}
```

### Domain-Driven-Design Folder Structure (Code Separation)

Regarding the DDD part of the refactoring, it would mean that the Readers and Senders and Handlers would be related to their Functionality based in the Game, rather than what they are (Senders/Readers).

Imagine the following folder structure:

```
- Communication/Messages
  - Character/
    - Session/
      - Reader.cs
      - Handler.cs
      - Sender.cs
    - Inventory
      - Reader.cs
      - Handler.cs
      - Sender.cs
    ...
  ...
```

This is an example of a possible Folder structure. The naming varies based on how we would separate and group the Messages.

#### An example of this strategy would be:

- Messages related to the Character should reside in the bigger Group "Character"
- Messages specific to the Character to manage its Inventory should reside inside "Character/Inventory"...
- Messages related to Managing the state/effects of Monsters should reside in the bigger Group "Monsters"
- Messages related to the Character to use Skills should reside inside "Character/Skills"

This is just an illustrative example.

**Comment by exectails** [comment]
Shouldn't this be a discussion for now? In any case, I've been going back and forth on this for 10 years with other developers, and I still always come back to the current design for its simplicity. Let's go over the problems I see.

First, there are the readers and writers/senders. I agree that, generally, it would make sense to put those into separate classes. It's clean, you only ever see that one packet, you can unit test them, but from a server emulator coder's perspective, it's a really annoying design. Every single time you want to add a packet handler, you have to go create one or two new files for the reader and the writer and implement them. Then you go and create the actual handler and potentially sender. Taking hopping between files into account, you spend what feels like 10x the amount of time on just adding the basic handling for a packet, especially because you often times don't get a packet right on the first try. And this adds up quickly, especially in these early stages, when you're constantly adding new packets. And then, whenever you need to make changes, you're hopping back and forth between files again. 

Adding categorization on top of of that makes it even worse, because then you're always asking yourself where you should put something, where you put things in the past, where things should go which's purpose you don't know yet, if maybe you should be moving something, etc.

In my eyes, this doesn't work well for a server emulator, where requirements constantly change and evolve and new discoveries change how packets are used and interpreted. It just adds additional steps, but offers very little benefit in my opinion. It lowers the file size of these classes a little, yes, but is that significant? When I need to find a packet I use the reference search and jump there immediately. The size of the overall file matters little, and I see the entire packet handling at a glance, which in most cases is exactly what I need in our use case. Organizing the packets "properly" also lets you unit test them, certainly a point, weren't it for the fact that packets are in constant flux and whether they're correct isn't determined by our unit tests, but the client. We aren't in control here at all.

Under normal circumstances I wouldn't use our current design. I would use something more or less like you proposed. But if I don't know what's coming and I constantly have to add, remove, modify, and tweak packets, putting everything in one place makes things much easier and straight forward in my opinion. 4000 lines of packet senders is a lot, yes. But personally, I think I'd rather split that class up into partial classes at some point, if actually necessary, instead of adding literally over 1000 separate files.

**Reply by ovflowd** [reply]
> Every single time you want to add a packet handler, you have to go create one or two new files

It will not be one file for the packet (handler and sender and reader). We would try to group packets by Category. That's what I was trying to explain. (Eg: Packets related to the Character itself (Movement, Naming, Effects, etc) go to the Character/ folder. Packet related to the Inventory of the Character goes to Character/Inventory... And it goes on. 

Edit: Just saw your comment about categories 🙈 

>  Then you go and create the actual handler and potentially sender. Taking hopping between files into account, you spend what feels like 10x the amount of time on just adding the basic handling for a packet, especially because you oftentimes don't get a packet right on the first try.

I can't agree with that, sorry. (But that's my personal opinion 🙈) I believe this depends personally from developer to developer.

> In my eyes, this doesn't work well for a server emulator, where requirements constantly change and evolve and new discoveries change how packets are used and interpreted.

As you said, in your eyes xD.

In all the Emulators that I know... (Bukkit, Kronus, Arcturus (Habbo), Holograph, and so many other games...) this approach is done. 

It's not only a design choice, it's an efficient design choice. Categorizing and making the structure of our Message Handling system clean and organised not only improves bug identification but also when we want to update the codebase.

> The size of the overall file matters little, and I see the entire packet handling at one glance

Hmm... I disagree. Finding packets only becomes easy in one file if you know exactly the Packet name. You can also with categorized/structured packets find them easily just by "Find by folder"...

For me a newcomer for Melia and ToS in general it's quite confusing which packet does what and where I can find similar packets since the packets are completely out of order (most of the time)

> But if I don't know what's coming and I constantly have to add, remove, modify, and tweak packets

Still categorizing is not a problem here. If you are working on a new packet that is about the Character, you'll just add it to the Character.ts file.

At least I can't really think how the Categorizing approach will slow down development... (Even with your examples, that I can't really imagine how they would actually be problematic 😅)

**Reply by ovflowd** [reply]
FYI: The folder structure I said before is just an example.

Readers, Senders and Handlers could even be in the same Class. For example Character Inventory packets could be something like `Communication/Messages/Character/Session.ts`

With something like:


```csharp
[PacketHandler(Op.CZ_LOGOUT)]
public void logoutUser(ChannelConnection conn)
{
    var { extra, unkByte } = this.CZ_LOGOUT();

    Log.Info("User '{0}' is logging out.", conn.Account.Name);

	Send(this.ZC_SAVE_INFO(conn));
	Send(this.ZC_LOGOUT_OK(conn));
}
```

Honestly, this is something that can be completely discussed 😄 (There are infinite approaches that are possible). This is how I did on a Habbo Hotel emulator: https://github.com/ovflowd/yupi/tree/linux/Yupi.Messages

**Reply by ovflowd** [reply]
What about something like this:

We separate group packets by "category", for example
```txt
- Handlers/
  - Character/
    - Inventory/
      - AddItemToInventoryMessageEvent.cs
  	  - DeleteItemFromInventoryMessageEvent.cs
  	- Session/
  	  - LogoutUserMessageEvent.cs
- Composers/
  - Character/
    - Session/
      - SaveUserInformationMessageComposer.cs
      - ConfirmUserLogoutMessageComposer.cs   
```

Does this structure make more sense?

**Reply by exectails** [reply]
>Categorizing and making the structure of our Message Handling system clean and organised not only improves bug identification but also when we want to update the codebase

Could you elaborate on this? How does it improve bug identification, or affect codebase updates in your opinion?

>In all the Emulators that I know... (Bukkit, Kronus, Arcturus (Habbo), Holograph, and so many other games...) this approach is done.

Of course, there are a lot of servers that use such a design. There's also a lot of servers that put the entire handling into dedicated files per packet, or servers where all packets are handled in one switch case. And there are servers who do it like we currently are. I don't mean to say that it's perfect, or that there's no better alternative, it's just been my experience that the current approach allows for fast and simple development, with close to no objective disadvantages as far as I could see.

> Finding packets only becomes easy in one file if you know exactly the Packet name. You can also with categorized/structured packets find them easily just by "Find by folder"

You're right, if you have absolutely no idea what you're looking for, randomly scrolling through the handler class won't exactly help either. And there's a case to be made for being able to find all inventory packets in some kind of inventory folder structure. However, I wonder how big of a factor that actually is. Aside from huge files being kind of intimidating, do you believe there will be many cases where someone opens Melia, has no idea what the name of the packet is, but needs to find a specific handler? You don't usually search at random, you might need to update a packet's structure, in which case you'll know the name. Or you might want to see the structure of a sent packet, in which case you might be coming from, say, the inventory class, where the sender is called. Off the top of my head, I can't think of many reasons why you would search for all inventory packets. What is the case you're proposing, where someone needs to find related packets?

>Even with your examples, that I can't really imagine how they would actually be problematic

Honest question, have you tried both approaches for a little while before? Across packet updates, on a project in its early stages, on a game that is in active development? Naturally there is some subjectivity to it. Something that bothers me, like having to hopp between multiple files until I fully grasp a packet (in some designs), or having to think about where to put certain packets that kind of fit nowhere, when all I want is to quickly throw together a dummy handler for something, so the client doesn't crash anymore, doesn't have to bother everyone. Personally, I can work more efficiently with this design, though it's also important to accomodate potential contributors. Though I also can't count the number of times contributors asked me where to put something, in projects that did use categories. Because that's not always straight forward either.

Edit: Sorry, I keep accidentally posting it as a new message xD

**Reply by ovflowd** [reply]
> Of course, there are a lot of servers that use such a design. There's also a lot of servers that put the entire handling into dedicated files per packet, or servers where all packets are handled in one switch case. And there are servers who do it like we currently are. I don't mean to say that it's perfect, or that there's no better alternative, it's just been my experience that the current approach allows for fast and simple development, with close to no objective disadvantages as far as I could see.

Yeah, all approaches are valid. I have personally the feeling that a more organised approach would be better.

> Could you elaborate on this? How does it improve bug identification, or affect codebase updates in your opinion?

I honestly have no technical or actual explanation. My answer was "gut-feeling" based. Disregard it 🙈 

>  Aside from huge files being kind of intimidating, do you believe there will be many cases where someone opens Melia, has no idea what the name of the packet is, but needs to find a specific handler?

My experience in Emulation is that many people from the Community like to touch the source, just because they can, and sometimes people like to contribute even with the smallest thing. But usually, huge files are intimidating, that's basically the biggest factor for newcomers. Also, not everyone knows what they're searching for, even when they're veterans in software development...

> Off the top of my head, I can't think of many reasons why you would search for all inventory packets. What is the case you're proposing, where someone needs to find related packets?

I believe it's more to keep it organised, it's more like a "practical" reason, like why Libraries keep books separated by categories? I'm not necessarily searching for something specific, but when I open a specific "Inventory.cs" and see something that might be interesting, it helps a lot 🤔 (My example is bad)

> Honest question, have you tried both approaches for a little while before? Across packet updates, on a project in its early stages, on a game that is in active development? Naturally, there is some subjectivity to it.

Actually yes. There was an Emulator for Habbo Hotel (Plus Emulator) that had the All Packets one-file approach, and it was a nightmare to maintain... But this is my personal opinion. So I'm definitely biased. And I apologise for that 😅 

>  Something that bothers me, like having to hopp between multiple files until I fully grasp a packet (in some designs), or having to think about where to put certain packets that kind of fit nowhere, when all I want is to quickly throw together a dummy handler for something, so the client doesn't crash anymore, doesn't have to bother everyone.

Usually, when I needed to test new packets, I also would just add a "Temporary.cs" file, just to test packets and their structures, and when I refactor it and have a clear goal and vision of what is their purpose and for what I would use them, I would then move them to an actual/proper folder/file. And then commit 😛 

> . Though I also can't count the number of times contributors asked me where to put something, in projects that did use categories. Because that's not always straight forward either.

That's true. That's why I believe that, if we categorise them, we should find a common denominator of how to "categorize" them. There are also always Misc Packets that don't fit anywhere... So just "Misc.cs" xD

**Reply by exectails** [reply]
>that had the All Packets one-file approach, and it was a nightmare to maintain

Tell me more, I'm curious what the problems were.

To me it's all about the pros and cons. I can tell you 5 reasons why I like the current design, 1 why I dislike it, 5 reasons I dislike a more organized approach, and maybe 2 why I like it. (More or less.) Sum it up and the current approach wins for me personally^^

Edit: Instead of "like", maybe a better wording is "why I think it has an advantage".

**Reply by ovflowd** [reply]
> Tell me more, I'm curious what the problems were.

I only remember vaguely, since it was more than 5 years ago. But I do remember that finding a specific message (also because of the naming of the methods) was kinda difficult. Mostly that.

But honestly, if you feel that in a long run the current approach suits us, sure. I don't really mind. I believe the names of the methods right now have the ID of the packets, so to find something specific if I have the packet in mind is easy.

What I believe could be done then, is in the Wiki a general description of what each Handler and Sender does.

Eg.: `CZ_LOGOUT` - Handles the Request from the Client that a user desires to Logout.

**Reply by exectails** [reply]
Such a description is supposed to be part of the XML documentation comments of each handler function, which we could easily generate documentation out of. Assuming that we document the code properly.

Btw, I want to mention that it's not my intention to dismiss suggestions, I simply state my opinion and what my past experiences were. Unfortunately such discussions rarely go anywhere with only 2-3 people involved, but generally I think we should have good reasons to make big changes to the overall structure.

**Reply by ovflowd** [reply]
> Such a description is supposed to be part of the XML documentation comments of each handler function, which we could easily generate documentation out of. Assuming that we document the code properly.

Yeah, I saw they had XML Docs. We could indeed generate some nice docs out of that 👍 

> Unfortunately such discussions rarely go anywhere with only 2-3 people involved, but generally, I think we should have good reasons to make big changes to the overall structure.

I completely agree! Let's see if the other folks want to give their opinions 🤔 

> Such a description is supposed to be part of the XML documentation comments of each handler function, which we could easily generate documentation out of. Assuming that we document the code properly.

You're not being dismissive at all! No worries 😄, I also don't really have counter-examples for this case, honestly. Because this is not a technological decision, but more a cosmetic one.

**Reply by Adephx** [reply]
Sorry, I have nothing worth mentioning to contribute to the discussion. I however always support classifications/categorizations. Also, if people can handle class names in Java, then perhaps we can afford to make our naming less arbitrary and for example change `ZC_LOGOUT_OK` to `ZONE_SRV_TO_CLIENT_LOGOUT_CONFIRMATION`, which should go a long way of distinguishing between different packages.

**Reply by ovflowd** [reply]
Agreed with @Adephx I believe whilst it definitely makes sense to decorate the methods with their respective Packets. `[PacketHandler(Op.CZ_LOGOUT)]` I would possibly use more human readable and camel-cased method names. But that's just yet another convention. So the handler of `CZ_LOGOUT` would be `handleUserLogout` and `ZC_LOGOUT_OK` would be `confirmUserLogout` (For example)

**Reply by exectails** [reply]
I understand where you're coming from, and under other circumstances I wouldn't use abbreviated, upper case names for handlers and senders either, but don't you think for developing a server emulator it's more important to be able to quickly and accurately find the packets and opcodes you're looking for? Because that's going to be a big portion of what we'll be doing, for months and years to come, until we have a solid grasp of the game's entire communication protocol.

If you were to use other names, you wouldn't be able to use a symbol search anymore and you would not be able to tell which packet it is just from the name of the method. Those things would be highly inconvienent. A symbol search would only lead you to the opcode, which you would then have to run a reference search on, or you'd have to do a normal search and click through the results to find the one you're looking for. And for senders it would kind of obfuscate what they're doing. Every time you see `Send.Xyz`, you would have to go into that method to see the actual opcode. But that op is really important, because you don't want to "confirm a user logout", you _need_ to send `ZC_LOGOUT_OK`, because that's what the client expects. And another name doesn't communicate that, you'd need to know the inner workings of `ConfirmUserLogout` to use it accurately.

I've heard the argument before, that you can rely on Intellisense for that, that you can just hover over it and see additional information, but that doesn't help if you aren't using a big IDE for once, or if you're reviewing commits and PRs. You could go and just rename all opcodes to something that you like better, but that's inconvenient because we're using a list of the official opcodes, which we would then have to "translate" regularly, and there's always going to be packets which's purpose we don't fully understand, making naming difficult.

**Reply by Adephx** [reply]
Sorry, there might be some misunderstanding on my part. Does the client expect the literal `ZC_LOGOUT_OK` string or in this case a `0xBC3` OPCode? Was `ZC_LOGOUT_OK` named by you or by IMC Games? If it's the latter, then I take everything back and agree that we should leave it as it is.

**Reply by exectails** [reply]
The client doesn't expect the literal string "ZC_LOGOUT_OK", but `ZC_LOGOUT_OK` _is_ the official name of the packet. The entire opcode list is official and is extracted by me whenever the ops change.

**Reply by ovflowd** [reply]
> you think for developing a server emulator it's more important to be able to quickly and accurately find the packets and opcodes you're looking for?

Isn't why I said that the methods would still be decorated with `[PacketHandler(Op.CZ_LOGOUT)]` for example?

So the actual method names could be cleaner? 🤔 (You can still nicely and easily search by the OpCodes since the methods are decorated by them)

**Reply by Adephx** [reply]
Alright, if `ZC_LOGOUT_OK` is a value that is part of the `0xBC3` package, then we should leave the entire list as it is.

**Reply by exectails** [reply]
>You can still nicely and easily search by the OpCodes since the methods are decorated by them

You can't use a symbol search though. With that, I copy the name of the packet I'm looking for, like from Pale, hit 3 keys in VS, and I'm there. If I have to do a regular search, I need 3 key hits to even start the search, depending on the size of the source code that search might take a few seconds, and then I'm greeted with the results, which I have to parse first, before getting to where I want to go, because they include all instances of that search string in the entire source code, including comments.

![](https://i.imgur.com/wsAL1aP.png)

Symbol searches are increadibly convenient, because you're searching for what you're actually looking for, a symbol that contains a certain string.

![](https://i.imgur.com/hmMTQ39.png)

Naturally, this isn't a requirement, but it's a big creature comfort when working with packets at this stage in such a project, because you constantly need to get to specific handlers and senders, and the time those more involved searches take adds up over time. I don't know if you're using symbol searches yourself, but I'm dreading every time I have to search for a packet in some of my bigger C++ projects, where the IDE has trouble properly indexing the whole thing, and I have to resort to normal searches, because the symbol search isn't reliable there.

**Comment by exectails** [comment]
I'm going to close this discussion for the time being, because it's my opinion that our current approach, while arguably a little messy at times, allows for much faster iteration than categorized approaches (see my points laid out above). This is more important to me personally at this point in the development process, though we may come back to this once our packet handling has reached a higher maturity level.
